File: Practice/AOJ/GRL/GRL2_A/grl2_a.py
# クエリを与えられるタイプのクラスカル
# クラスカル自体はAlDS1_12Aでやったのでほとんど流用するだけ

n, m = map(int, input().split())
graph = []


class Edge:
  def __init__(self, u, v, cost):
    self.u = u
    self.v = v
    self.cost = cost


parent = [i for i in range(n)]


def root(x):
  if x == parent[x]:
    return x
  else:
    parent[x] = root(parent[x])
    return parent[x]


# 最小のコスト順からやるっていうからには
# いったんすべてを取らねばならない....
for _ in range(m):
  u, v, cost = map(int, input().split())
  e = Edge(u, v, cost)
  graph.append(e)

graph = sorted(graph, key=lambda g: g.cost)

# そんで探索
ans = 0
for e in graph:
  u, v = root(e.u), root(e.v)
  if u != v:
    parent[v] = u
    ans += e.cost

print(ans)

File: Practice/AOJ/GRL/GRL2_A/_grl2_a.go
// ふつうにソートでやるとTLEなった これはPriority Queueで叩くやつだった

package main

import (
	"fmt"
	"sort"
)

type Edge struct {
	x, y, cost int
}

var parents []int
var edges []Edge

func main() {
	var v, e int
	fmt.Scan(&v, &e)
	parents = make([]int, v)
	edges = make([]Edge, e)
	for i := 0; i < v; i++ {
		parents[i] = i
	}
	for i := 0; i < e; i++ {
		var s, t, w int
		fmt.Scan(&s, &t, &w)
		edges[i] = Edge{x: s, y: t, cost: w}
	}
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].cost < edges[j].cost
	})
	ans := 0
	for _, e := range edges {
		if !same(e.x, e.y) {
			union(e.x, e.y)
			ans += e.cost
		}
	}
	fmt.Println(ans)
}

func root(x int) int {
	if x == parents[x] {
		return x
	}
	y := root(parents[x])
	parents[x] = y
	return y
}

func same(a, b int) bool {
	return root(a) == root(b)
}

func union(a, b int) {
	ra, rb := root(a), root(b)
	parents[ra] = rb
}
